Gutenboarding: Customize steps for Anchor.fm

allilevine opened the pull request:
#### Changes proposed in this Pull Request

* Store URL query parameter and change steps based on it
* Prevent site creation immediately after signup -- show steps instead
* Prevent plans and features steps from showing up
* All changes are behind the `anchor-fm-dev` feature flag
  * Flag is currently set on in development, off in other environments

#### Testing instructions

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->

* Go to http://calypso.localhost:3000/new?anchor_podcast=40a166a8
* Create a new account or log in
* Examine the updated copy to name your podcast
- [ ] You can specify the site title, design, and fonts, and then set up a site.

#### Missing

- Current designs shown are two placeholder copies.

#### To do
- [x] Show steps after login. The login [redirects back to `Step.CreateSite`](https://github.com/Automattic/wp-calypso/blob/691cc7cc63da06db6a0a13204ed2e005cfc4990e/client/landing/gutenboarding/components/signup-form/index.tsx#L162). We should redirect it to the intent gathering step instead.
- [x] Persist user data after initial signup (I'm asked to sign up again after the steps).
- [x] Sanitize URL query parameters.
- [x] Trigger signup with podcast id, not extra signup parameter. 
- [x] Change signup and login forms with custom header content.
- [x] Pass the podcast id to [the site creation action](https://github.com/Automattic/wp-calypso/blob/691cc7cc63da06db6a0a13204ed2e005cfc4990e/client/landing/gutenboarding/stores/onboard/actions.ts#L63). (@mreishus)
- [x] Change intent gathering text.
<img width="652" alt="Screen Shot 2020-12-01 at 1 32 59 AM" src="https://user-images.githubusercontent.com/66652282/100705416-27fcc800-3375-11eb-85bf-56b3ad98048c.png">

- [x] Update Choose Design copy.
<img width="809" alt="Screen Shot 2020-12-01 at 1 28 08 AM" src="https://user-images.githubusercontent.com/66652282/100705548-57abd000-3375-11eb-9a23-8bba225d8468.png">

- [x] Update  number of columns for design available.
<img width="1363" alt="Screen Shot 2020-12-01 at 1 26 55 AM" src="https://user-images.githubusercontent.com/66652282/100705570-672b1900-3375-11eb-9bc2-fa54a145c535.png">

- [x] Add ability to limit available designs.
<img width="1491" alt="Screen Shot 2020-12-01 at 5 41 47 PM" src="https://user-images.githubusercontent.com/66652282/100805529-ce3fe080-33fc-11eb-8b90-321d4f83c766.png">

#### Issues
- [x] F5 after signup (direct) (Fixed in fb724b49d)
  - http://calypso.localhost:3000/new?anchor_podcast=2&signup
  - Insert new email, password, click create account
  - See the "My site is named page"
  - Press F5
  - Now you see the registration screen again. I expected to remain logged in with my new account.
  - Note: This might change when `Trigger signup with podcast id, not extra signup parameter.` is implemented.
- [x] F5 after signup (delayed) (Fixed in 34ef59a2cb1)
  - In the use-step-navigation file, console.log the value of `isAnchorFmSignup`
  - http://calypso.localhost:3000/new?anchor_podcast=2&signup
  - Insert new email, password, click create account
  - Fill out a site name, click continue
  - Now on choose a design page. Open console, note that `isAnchorFmSignup` is true
  - F5.  Same page loads, but in console, `isAnchorFmSignup` is now false.

allilevine committed:
Check for anchor_podcast query param and change steps if present

allilevine committed:
Clarify conditions with comments

allilevine committed:
Show steps after signup

allilevine committed:
Change signup form for podcast signups

allilevine committed:
Store podcastId in automattic/onboard

mreishus commented on client/landing/gutenboarding/components/signup-form/index.tsx:
```diff
@@ -43,6 +43,7 @@ const SignupForm = ( { onRequestClose }: Props ) => {
 	const makePath = usePath();
 	const currentStep = useCurrentStep();
 	const isMobile = useViewportMatch( 'small', '<' );
+	const isAnchorFmSignup = useSelect( ( select ) => select( ONBOARD_STORE ).getSelectedPodcastId() );
```
Noting: isAnchorFmSignup sounds like a boolean, but it's a string with value `"2"` (when using Anchor), or `""` (empty string, when not using anchor). It does work, since empty string is a falsey value, it just wasn't what I expected.

getdave replied:
I agree. `!!` or `Boolean()` to make it explicit as a bool

mreishus replied:
Added boolean selector here ac1e119fb2

getdave commented on client/landing/gutenboarding/hooks/use-on-login.ts:
```diff
@@ -27,9 +27,10 @@ export default function useOnSignup() {
 
 	React.useEffect( () => {
-		if ( ! isCreatingSite && ! newSite && currentUser && shouldTriggerCreate ) {
+		if ( ! isCreatingSite && ! newSite && currentUser && shouldTriggerCreate && ! isAnchorFmSignup ) {
```
`isAnchorFmSignup` is missing from `useEffect` deps.

mreishus replied:
Filled out deps arrays here 30b0272d3c

getdave commented on client/landing/gutenboarding/hooks/use-on-signup.ts:
```diff
@@ -24,6 +25,12 @@ export default function useOnSignup() {
+
+	React.useEffect( () => {
+		dispatch( ONBOARD_STORE ).setPodcastId( anchorFmPodcastId );
+	}, [] );
```
Do we want this to run only once? What if `isAnchorFmSignup` changes?

mreishus replied:
Filled out deps arrays here 30b0272d3c

mreishus replied:
I had to undo the change here: e5ebad75a3 .  The anchor sign up process no longer works after this is set to update whenever `anchorFmPodcastId` changes.  The reason being, the query parameter is only available on the initial pageload, and is lost as we navigate between steps.  

Not 100% sure about this; maybe it's better to keep it responsive and to modify use-step-navigation so it keeps the anchor query param around..

getdave commented on client/landing/gutenboarding/hooks/use-on-signup.ts:
```diff
@@ -33,7 +40,7 @@ export default function useOnSignup() {
 
 	React.useEffect( () => {
-		if ( newUser && newUser.bearerToken && newUser.username && ! newSite ) {
+		if ( newUser && newUser.bearerToken && newUser.username && ! newSite && ! isAnchorFmSignup ) {
```
`isAnchorFmSignup` missing hook dependency.

mreishus replied:
Filled out deps arrays here 30b0272d3c

getdave commented on client/landing/gutenboarding/hooks/use-on-signup.ts:
```diff
@@ -24,6 +25,12 @@ export default function useOnSignup() {
+	const isAnchorFmSignup = useSelect( ( select ) => select( ONBOARD_STORE ).getSelectedPodcastId() );
+
+	React.useEffect( () => {
+		dispatch( ONBOARD_STORE ).setPodcastId( anchorFmPodcastId );
```
`anchorFmPodcastId` does this come direct from URL query params? Do we need/want to santize it?

allilevine replied:
It does! Sanitizing it sounds like a good idea. 👍

allilevine replied:
I added [a check for any non-numeric characters](https://github.com/Automattic/wp-calypso/pull/47600/commits/ed32e476902b24339280e308c0a23411ffe3e7c4).

mreishus committed:
Run prettier; formatting changes only

mreishus committed:
Fill out hook dep arrays

mreishus committed:
One more prettier run

mreishus committed:
Create new isAnchorFmSignup selector, boolean

mreishus committed:
Unfill dep array

This breaks the signup process because the ?anchor_podcast=2 query param
doesn't get passed from step to step, it only exists on the first step.

mreishus committed:
Anchor.fm onboarding; Don't ask for login information twice

http://calypso.localhost:3000/new?anchor_podcast=2&signup

allilevine committed:
Add newUser.bearerToken to createSite call

mreishus commented:
We're seeing some weird things in dev testing.  In some cases, during site creation, you're asked to log in a second time with an error about no access to private blog in the browser console.

Chromium does not appear to work, but firefox does. (Possibly related: #46127).  Using `@example.com` email addresses works, but work emails with `+` signs in them don't.

allilevine committed:
Revert adding newUser.bearerToken

allilevine committed:
Redirect to intent gathering on login

allilevine committed:
Rename podcastId to anchorFmPodcastId to be consistent.

allilevine committed:
Validate the podcast id

allilevine committed:
Change signup subtitle for podcast signups

allilevine committed:
Fix typo

mreishus committed:
Merge branch 'trunk' into add/custom-anchor-gutenboarding-steps

mreishus committed:
Run prettier

mreishus commented:
Noted some F5 problems in the PR description at the top.  These can possibly be handed by using `state` in our `history.push`, and using `useLocation` to read out that state later.  I started coding this, but it got a little bit too complex imo, so I created this refactor: #47882.  It will conflict with this branch, but I can resolve that np.

krymson24 commented:
I'm not sure if I missed something, but did we opt to not make the change for the "My site is called" copy to "My podcast is called"?

I can make the change here if we still require it:
```client/landing/gutenboarding/onboarding-block/acquire-intent/site-title.tsx`

krymson24 committed:
add(gutenboarding) add input label to gutenboarding

krymson24 committed:
add(gutenboarding) update choose design copy for podcast sites

krymson24 commented:
I made some updates to this PR based on the changes it looks like we needed.

I left off at limiting the available designs -- there's a TODO comment in `client/landing/gutenboarding/index.tsx`

mreishus committed:
Merge branch 'trunk' into add/custom-anchor-gutenboarding-steps

obenland commented on client/landing/gutenboarding/components/signup-form/index.tsx:
```diff
@@ -158,16 +164,25 @@ const SignupForm = ( { onRequestClose }: Props ) => {
 		<Modal
 			className={ 'signup-form' }
-			title={ __( 'Save your progress' ) }
+			title={ __( modalTitle ) }
```
In PHP those translation functions can't process variables, is that different in JS?
If it's not, feel free to just translate all strings in the ternary above and pass the translated string in the variable.

allilevine replied:
[It's the same](https://github.com/Automattic/i18n-calypso/blob/HEAD/README.md#strings-only), thanks for pointing that out! I [translated the strings in the ternary](https://github.com/Automattic/wp-calypso/pull/47600/commits/a0f9e317771ff52ccc3ce7f6fa2f7c7ecaad1f92) instead.

mreishus committed:
Merge branch 'trunk' into add/custom-anchor-gutenboarding-steps

mreishus commented:
My small refactor is merged into trunk, then this branch with conflicts resolved.  `client/landing/gutenboarding/hooks/use-steps.ts` creates a list of steps only, and `client/landing/gutenboarding/hooks/use-step-navigation.ts` uses those steps to handle site creation and the forward/back actions.

allilevine committed:
Pass strings to i18n

krymson24 committed:
add(gutenboarding) two columns for available designs

krymson24 committed:
add(gutenboarding) add ability to limit designs for anchorfm site templates

krymson24 commented:
Finished
- Change intent gathering text
- Update Choose Design copy
- Update number of columns used
- Allow for filtering designs only for podcasting sites

Left an additional `TODO` to add the finalized design templates, and add the corresponding `is_anchorfm` prop to those templates.

mreishus committed:
Run prettier

mreishus commented:
For regular gutenboarding, pressing F5 does not remove state.  For example, you might choose font #3, move to another step, press F5, and the page still remembers that you have selected that font, even having reloading.  The redux state is persisted.

For this branch, pressing F5 wipes out the anchor podcast id sometimes.  That's because:
  - We migrate the podcast id from the query parameter to redux
  - Clicking next/previous steps removes the query parameter from the url
  - Pressing F5 causes the podcast id to be re-migrated from the query parameter to redux, so even though the redux key is preserved, this re-migration effect hook wipes it out.

I have a prospective solution in a PR that merges into this branch here: #47941. It removes the anchor podcast ID from redux and stores it in "location state" handled by react-router, and adds that state to be passed on when clicking the next/previous buttons.  It fixes the F5 problem in my testing.

Possibly there's a way to solve the F5 problem while also storing the value in redux and continuing to migrate it from the query param, but I didn't find one.  Open to alternate suggestions or reviews.

krymson24 commented on client/landing/gutenboarding/onboarding-block/design-selector/index.tsx:
```diff
@@ -46,61 +50,74 @@ const DesignSelector: React.FunctionComponent = () => {
+					className={ isAnchorFmSignup ? 'design-selector__grid-minimal' : 'design-selector__grid' }
+				>
+					{ getRandomizedDesigns()
+						.featured.filter(
```
Despite the suggested location for where to add this filter in `client/landing/gutenboarding/available-designs.ts`, podcast sites are not determined via the Config API.

Available designs also get stored on app bootstrap in `client/landing/gutenboarding/index.tsx`, which is also not an appropriate place to add this, since the design step could be in any part of the site creation flow.

Lastly, the filtering is done here so that the available designs remain as the source of truth, and the filtering is done on the view, right before being shown to the user.

mreishus committed:
Merge branch 'trunk' into add/custom-anchor-gutenboarding-steps

mreishus committed:
Move the "anchor fm podcast id" state from redux to "location state" (#47941)

* Move the "anchor fm podcast id" state from redux to "location state" provided by react-router

- Fixes F5 problem, where pressing F5 on a step moves you to the
  non-anchor flow

* Inline useAnchorFmQueryParam()

* Rename state to locationState

mreishus committed:
design-selector: Show designs without is_anchorfm when in non-anchor flow

mreishus committed:
Add two placeholder designs with is_anchorfm set to true

This way, we can see some designs to pick while testing the anchor flow.
They're copies of existing designs, and not final.

allilevine committed:
Trigger signup if we have a podcast id but no user

mreishus commented:
Working on this one: "Pass the podcast id to the site creation action.".  Submitted a refactor request to trunk in #47980 as a first step.

allilevine committed:
Sanitize podcast id

mreishus committed:
Merge branch 'trunk' into add/custom-anchor-gutenboarding-steps

mreishus committed:
Run prettier

mreishus committed:
Merge branch 'trunk' into add/custom-anchor-gutenboarding-steps

mreishus committed:
Remove comment about testing non-anchor gutenboarding

I've tested it and it works.

mreishus committed:
Gutenboarding Anchor: anchor_fm_podcast_id makes it to wpcom backend

mreishus commented:
I have the podcast ID flowing through the create site Ajax request:  
![2020-12-03_14-30 (Create Anchor Redacted)](https://user-images.githubusercontent.com/937354/101091152-66fd6a00-357d-11eb-8af0-bdaefbc3af5d.png)
(Starts in initial onboarding query param, is stored in location state, then passed to the createSite call)

mreishus committed:
Merge branch 'trunk' into add/custom-anchor-gutenboarding-steps

mreishus committed:
Put Anchor gutenboarding behind anchor-fm-dev feature flag

mreishus commented:
## Looking for Review

We're in at a spot where we're looking for another round of reviews, aiming to merge this code to trunk while inactive, behind the `anchor-fm-dev` feature flag, which is only active in dev.  The goal of _deploying_ the code, but not _releasing_ it is to:
  - Reduce the size of future PRs (they're easier to review and work on in iterations)
  - Reduce the 'surface area' for future conflicts

## What's there

Right now, you can:
  - Provide a query parameter to get the anchor-specific flow, like: http://calypso.localhost:3000/new?anchor_podcast=40a166a8
  - See updated wording, now referring to podcasts
  - See the plan and feature steps removed
  - See site creation at the end, with login/signup at the beginning
    - It's okay to create new accounts+sites for testing, I've been using emails with `@example.com` in them
  - Use network dev tools to see the call to `/new` includes the `anchor_fm_podcast_id` under `options` (See screenshot above)

## What's missing

We're still missing:
  - The anchor specific designs (Two placeholder designs are present)
  - Coordination with wpcom (An extra parameter is sent to wpcom site creation, but the php code doesn't read it)
  - Possibly Episode GUID integration #350
  - Fetching podcast name #351

krymson24 commented on client/landing/gutenboarding/available-designs-config.json:
```diff
@@ -11,7 +11,8 @@
 			"is_premium": false,
-			"is_fse": true
+			"is_fse": true,
+			"is_anchorfm": false
```
Do we need to add this prop there or is it to be more explicit?

simison replied:
There's no `is_fse: false` neither in many of the designs.

Maybe these would be better as an array `flows: [ 'fse', 'anchor', ... ]` ?

simison replied:
To elaborate: actually, instead of thinking of these as "flows", it might be better to think of these as "supported features". Imagine we'd have adjustments in the flow for blog, e-commerce, business address, podcasting, or any combination of those.

mreishus replied:
I generally prefer being explicit over not.  For example, the current implementation of optional parameter `is_anchorfm` leads to this expression: 

`isAnchorFmSignup === ( design.is_anchorfm ?? false )`.

I'd prefer `isAnchorFmSignup == design.is_anchorfm`, but lint rules guide me towards using triple equals, and `false === undefined` isn't true.  It's easier to only have to think about one possible value (`false`) rather than two (`false` or `undefined`).

mreishus replied:
Created PR:  _Move is_anchor attribute on gutenboard designs to feature array #48062_

Note this only concerns itself with the anchor feature for now, to limit scope.  We can add other features after this is merged.

krymson24 replied:
My only concern is maintainability -- every new design requires this prop to be added. I am generally on the side of things being more explicit. My comment is not a blocker by any means!

mreishus replied:
Please see PR #48062

simison replied:
@mreishus while being explicit, will this then fail loudly if someone doesn't add the key in all designs as well _doesn't_ manually test Anchor flows? Note also that some day these hopefully will come from an API.

simison replied:
Oh and my primary concerns was situation where we have dozen `is_` things here and array felt like more scalable when we already/only have 3. :-)

Definetely not a blocker from my perspective tho 👍

mreishus replied:
* The `is_anchor` key is still optional, I only added extra false values to be explicit.
* I am not quite understanding the second point: why a new design without anchor set would require manual testing of Anchor flows.  I don't believe it should.
* PR #48602 removes the `is_anchor` attribute for a new features array.

krymson24 commented on client/landing/gutenboarding/components/signup-form/index.tsx:
```diff
@@ -158,16 +167,25 @@ const SignupForm = ( { onRequestClose }: Props ) => {
-		`${ window.location.origin }/new${ makePath( Step.CreateSite ) }?new`
+		`${ window.location.origin }/new${
+			isAnchorFmSignup ? makePath( Step.IntentGathering ) : makePath( Step.CreateSite )
+		}?new${ anchorFmParameter }`
```
So this could end up as `/new/path/?new`? Hmm..

allilevine replied:
Thanks, I [pushed a change](https://github.com/Automattic/wp-calypso/pull/47600/commits/38f5e4b52de57f9f5497447622779d0ec49c5a26) so that the two login redirect URL possibilities are:
- /new/create-site?new (the current redirect path to site creation)
- /new/?new&anchor_podcast={id} (the Anchor.fm flavored redirect path to the My podcast is called... step which is the root path)

krymson24 commented on client/landing/gutenboarding/components/signup-form/index.tsx:
```diff
@@ -158,16 +167,25 @@ const SignupForm = ( { onRequestClose }: Props ) => {
 		<Modal
 			className={ 'signup-form' }
-			title={ __( 'Save your progress' ) }
+			title={ modalTitle }
```
I thought I read somewhere that these couldn't be translated dynamically? If I'm mistaken, please ignore

allilevine replied:
I moved the translations into the ternary based on https://github.com/Automattic/wp-calypso/pull/47600#discussion_r533521894

The `modalTitle` is used twice so a variable is convenient, but maybe it would be better to just eliminate them altogether, [based on these examples](https://github.com/Automattic/i18n-calypso/blob/HEAD/README.md).

allilevine replied:
I [removed the variables](https://github.com/Automattic/wp-calypso/pull/47600/commits/56edf58cc8298035da7ed43f8d8db2c0904b156b).

krymson24 replied:
100% agree with you to keep this dry -- I just wasn't sure if someone mentioned that having a translated variable would work as expected

krymson24 commented on client/landing/gutenboarding/onboarding-block/design-selector/index.tsx:
```diff
@@ -46,61 +48,74 @@ const DesignSelector: React.FunctionComponent = () => {
+							( design ) =>
+								// TODO Add finalized design templates to client/landing/gutenboarding/available-designs-config.json
+								// along with is_anchorfm prop
+								isAnchorFmSignup === ( design.is_anchorfm ?? false )
```
What's the difference between this and `isAnchorFmSignup && design.is_anchorfm`?

mreishus replied:
Consider the case when `isAnchorFmSignup` is false, and `design.is_anchorfm` is false or undefined.

* `isAnchorFmSignup && design.is_anchorfm` evaluates to false.
  * Thus, in the non-anchor flow, no designs are shown.
* `isAnchorFmSignup === ( design.is_anchorfm ?? false )` evaluates to true.
  * Thus, in the non-anchor flow, all of the non-anchor designs are shown.

krymson24 replied:
You're right -- I thought about this just as I fell asleep 😴

krymson24 commented:
> ## Looking for Review
> We're in at a spot where we're looking for another round of reviews, aiming to merge this code to trunk while inactive, behind the `anchor-fm-dev` feature flag, which is only active in dev. The goal of _deploying_ the code, but not _releasing_ it is to:
> 
> * Reduce the size of future PRs (they're easier to review and work on in iterations)
> * Reduce the 'surface area' for future conflicts
> 
> ## What's there
> Right now, you can:
> 
> * Provide a query parameter to get the anchor-specific flow, like: http://calypso.localhost:3000/new?anchor_podcast=40a166a8
> * See updated wording, now referring to podcasts
> * See the plan and feature steps removed
> * See site creation at the end, with login/signup at the beginning
>   
>   * It's okay to create new accounts+sites for testing, I've been using emails with `@example.com` in them
> * Use network dev tools to see the call to `/new` includes the `anchor_fm_podcast_id` under `options` (See screenshot above)
> 
> ## What's missing
> We're still missing:
> 
> * The anchor specific designs (Two placeholder designs are present)
> * Coordination with wpcom (An extra parameter is sent to wpcom site creation, but the php code doesn't read it)
> * Possibly Episode GUID integration #350
> * Fetching podcast name #351

@mreishus thanks for this summary!

I gave this PR a set of eyes. Thank you for your refactor, and I appreciate the decision to move the `isAnchorFm` determination out of `Redux` to instead depend on `location`. This empowers us for stateless navigation and protects us against potential race conditions 👍 . I'll do some testing and more review tomorrow, if this is isn't merged yet.

allilevine committed:
Refactor login redirect logic

mreishus committed:
Merge branch 'trunk' into add/custom-anchor-gutenboarding-steps

mreishus committed:
Run prettier, no functional changes

allilevine committed:
Remove translated strings from variable

krymson24 commented on client/landing/gutenboarding/path.ts:
```diff
@@ -96,3 +104,32 @@ export function useCurrentStep() {
+// provided by react-router-dom, if not available there, checks the query string.
+export function useAnchorFmPodcastId(): string | null {
+	const { state: locationState, search } = useLocation< GutenLocationStateType >();
+	const sanitizePodcastId = ( id: string ) => id.replace( /[^a-zA-Z0-9]/g, '' );
```
Surprised the new URLSearchParams API doesn't already sanitize params for you. Also surprised that we don't have a helper function to do that?

alshakero replied:
I don't think we need to sanitize here.

allilevine replied:
> I don't think we need to sanitize here.

@alshakero Thanks! Would you mind clarifying why we don't need to sanitize here?

alshakero commented on client/landing/gutenboarding/hooks/use-on-login.ts:
```diff
@@ -27,14 +27,23 @@ export default function useOnLogin(): void {
 				languageSlug: locale,
 				bearerToken: undefined,
 				visibility,
+				anchorFmPodcastId,
```
Won't this always be `null` given that the if checks for `! isAnchorFmSignup`?

mreishus replied:
Sure, removed here: [add/custom-anchor-gutenboarding-steps 87646a6645] Gutenboarding: Remove useAnchorFmPodcastId() call from non-anchor siteCreate

alshakero commented on client/landing/gutenboarding/hooks/use-step-navigation.ts:
```diff
@@ -34,18 +40,34 @@ export default function useStepNavigation(): { goBack: () => void; goNext: () =>
+			} );
+		}
+		// Adding a newUser check works for Anchor.fm flow.  Without it, we ask for login twice.
+		if ( newUser && newUser.username !== undefined ) {
```
```suggestion
		if ( newUser?.username ) {
```

alshakero commented on client/landing/gutenboarding/path.ts:
```diff
@@ -96,3 +104,32 @@ export function useCurrentStep() {
+
+	// Fall back to looking in query param
+	const queryParam = new URLSearchParams( search ).get( 'anchor_podcast' );
+	if ( queryParam !== null && queryParam !== '' ) {
```
```suggestion
	if ( queryParam ) {
```

alshakero commented on client/landing/gutenboarding/path.ts:
```diff
@@ -96,3 +104,32 @@ export function useCurrentStep() {
 }
+
+export function useIsAnchorFm(): boolean {
+	const podcastId = useAnchorFmPodcastId();
+	return podcastId !== null && podcastId !== '';
```
```suggestion
	return Boolean(useAnchorFmPodcastId());
```

alshakero replied:
I'm a little worried that this might create some strange behaviors when people give wrong anchorFM ids. I'd check if they match a certain pattern. Because any non-empty string can pass this check and throw Gutenboarding flow off.

mreishus replied:
Added validation here: [add/custom-anchor-gutenboarding-steps 499dc09c4e] Validate podcast ID

alshakero commented on client/landing/gutenboarding/stores/onboard/actions.ts:
```diff
@@ -102,6 +104,9 @@ export function* createSite( {
 				visibility === Site.Visibility.PublicNotIndexed && {
 					wpcom_public_coming_soon: 1,
 				} ),
+			...( anchorFmPodcastId !== null && {
```
```suggestion
			...( anchorFmPodcastId && {
```

alshakero reviewed (changes requested):
Great work. 6/7 of the comments are minor and only one is a a bit unclear.

mreishus committed:
Merge branch 'trunk' into add/custom-anchor-gutenboarding-steps

mreishus committed:
Update client/landing/gutenboarding/stores/onboard/actions.ts

Co-authored-by: Omar Alshaker <omar@omaralshaker.com>

mreishus committed:
Update client/landing/gutenboarding/path.ts

Co-authored-by: Omar Alshaker <omar@omaralshaker.com>

mreishus committed:
Update client/landing/gutenboarding/hooks/use-step-navigation.ts

Co-authored-by: Omar Alshaker <omar@omaralshaker.com>

mreishus committed:
Gutenboarding: Remove useAnchorFmPodcastId() call from non-anchor siteCreate

mreishus committed:
Merge branch 'add/custom-anchor-gutenboarding-steps' of github.com:Automattic/wp-calypso into add/custom-anchor-gutenboarding-steps

mreishus committed:
Validate podcast ID

mreishus committed:
Anchor podcast IDs can be 7 or 8 chars long

mreishus commented:
re: e2e test failures.  My PR "Add white space #48094", is suffering the same failures, so they are nothing to worry about.

alshakero reviewed (approved):
Reviewed the code. Tested classic Gutenboarding and couldn't see any regression. Tested this PR and all works as expected. Thanks everyone!
